test/core/both/test-floodsub.js

```diff
@@ -0,0 +1,140 @@
+            nodeUnstarted = ipfs
+            done()
+          })
+        })
+      })
+    })
+  })
+
+  after((done) => {
+    factory.dismantle(() => done())
+  })
+
+  describe('Floodsub API', () => {
+    describe('start', () => {
+      it('throws if offline', () => {
+        expect(() => nodeOffline.floodsub.start()).to.throw
+      })
+
+      it('success', () => {
+        nodeStarted.floodsub.start((err, _floodsub) => {
+          expect(err).to.not.exist
+          expect(_floodsub).to.exist
+          floodsub = _floodsub
+        })
+      })
+    })
+
+    describe('subscribe', () => {
+      it('throws if offline', () => {
+        expect(() => nodeOffline.floodsub.subscribe()).to.throw
+      })
+
+      it('throws if not started', () => {
+        expect(() => nodeUnstarted.floodsub.subscribe(topic)).to.throw
+      })
+
+      it('throws without a topic', () => {
+        expect(() => nodeStarted.floodsub.subscribe()).to.throw
+      })
+
+      it('success', (done) => {
+        nodeStarted.floodsub.subscribe(topic, (err, stream) => {
+          expect(err).to.not.exist
+          expect(stream.readable).to.eql(true)
+          expect(typeof stream._read).to.eql('function')
+          expect(typeof stream.cancel).to.eql('function')
+          expect(stream instanceof Stream).to.eql(true)
+
+          subscription = stream
+
+          const nodeSubs = floodsub.getSubscriptions()
+          expect(nodeSubs.length).to.eql(1)
+          done()
+        })
+      })
+    })
+
+    describe('publish', () => {
+      it('throws if offline', () => {
+        expect(() => nodeOffline.floodsub.publish()).to.throw
+      })
+
+      it('throws if not started', () => {
+        expect(() => nodeUnstarted.floodsub.publish(topic, message)).to.throw
+      })
+
+      it('throws without a topic', () => {
+        expect(() => nodeStarted.floodsub.publish()).to.throw
+      })
+
+      it('throws without data', () => {
+        expect(() => nodeStarted.floodsub.publish(topic)).to.throw
+      })
+
+      it('success', (done) => {
+        const published = true
+
+        subscription.on('data', () => done())
+
+        nodeStarted.floodsub.publish(topic, message, () => {
+          expect(published).to.eql(true)
+        })
+      })
+    })
+
+    describe('unsubscribe', () => {
+      it('throws if offline', () => {
+        expect(() => nodeOffline.floodsub.unsubscribe()).to.throw
+      })
+
+      it('throws if not started', () => {
+        expect(() => nodeUnstarted.floodsub.unsubscribe(topic)).to.throw
+      })
+
+      it('throws without a topic', () => {
+        expect(() => nodeStarted.floodsub.unsubscribe()).to.throw
+      })
+
+      it('success', (done) => {
+        nodeStarted.floodsub.unsubscribe(topic, (err) => {
+          expect(err).to.not.exist
+          const nodeSubs = floodsub.getSubscriptions()
+          expect(nodeSubs.length).to.eql(0)
+          done()
+        })
+      })
+    })
+  })
+})
```

Comment:
It would be awesome to have these tests as part of interface-ipfs-core, so that we can use the same tests for js-ipfs and js-ipfs-api, making sure that both implementations follow the same interface. 

See https://github.com/ipfs/interface-ipfs-core/tree/master/src for examples in how to achieve this.